iso15118/secc/states/iso15118_2_states.py

```diff
@@ -577,61 +592,120 @@ async def process_message(
             V2GMessageV20,
             V2GMessageDINSPEC,
         ],
+        message_exi: bytes = None,
     ):
         msg = self.check_msg_v2(message, [CertificateInstallationReq])
         if not msg:
             return
 
-        cert_install_req: CertificateInstallationReq = (
-            msg.body.certificate_installation_req
+        if not self.validate_message_signature(msg):
+            self.stop_state_machine(
+                "Signature verification failed for " "CertificateInstallationReq",
+                message,
+                ResponseCode.FAILED_SIGNATURE_ERROR,
+            )
+
+        # In a real world scenario we need to fetch the certificate from the backend.
+        # We call get_15118_ev_certificate method, which would be a direct mapping
+        # to the Get15118EVCertificateRequest
+        # message from OCPP 2.0.1 for the installation case.
+        # This accepts 2 arguments:
+        # 1. The raw EXI CertificateInstallationReq message coming from the EV
+        # in base64 encoded form.
+        # 2. A string that specifies `15118SchemaVersion` which would be either of
+        # "urn:iso:15118:2:2013:MsgDef" or "urn:iso:std:iso:15118:-20:CommonMessages"
+        signature = None
+
+        try:
+            if self.comm_session.config.use_cpo_backend:
+                logger.info("Using CPO backend to fetch CertificateInstallationRes")
+                # CertificateInstallationReq must be base64 encoded before forwarding
+                # to backend.
+                # Call to b64encode returns byte[] - hence the .decode("utf-8")
+                base64_certificate_install_req = base64.b64encode(message_exi).decode(
+                    "utf-8"
+                )
+
+                # The response received below is EXI response in base64 encoded form.
+                # Decoding to EXI happens later just before V2GTP packet is built.
+                base64_certificate_installation_res = (
+                    await self.comm_session.evse_controller.get_15118_ev_certificate(
+                        base64_certificate_install_req, Namespace.ISO_V2_MSG_DEF
+                    )
+                )
+                certificate_installation_res: Base64 = Base64(
+                    payload=base64_certificate_installation_res,
+                    payload_type=CertificateInstallationRes.__name__,
```

Comment:
is this what we want? in fact why do we need that `payload_type` field? is not used anywhere and I was confusing it with the `next_msg_payload_type` from the `create_next_message` method

Reply:
This was added to help with logging in the following line. At the moment the only purpose of that field is to help understand what type of message this is when cast to str as is the behaviour with the other message types.

```
    def __str__(self):
        return self.message_type
``` 


https://github.com/SwitchEV/iso15118/blob/4e141ab2c9970f7d8286972cc79354b880303952/iso15118/shared/comm_session.py#L396

Reply:
maybe we can `s/payload_type/message_name`